src/Razor/src/Microsoft.CodeAnalysis.Remote.Razor/Hover/RemoteHoverService.cs

```diff
@@ -0,0 +1,213 @@
+{
+    internal sealed class Factory : FactoryBase<IRemoteHoverService>
+    {
+        protected override IRemoteHoverService CreateService(in ServiceArgs args)
+            => new RemoteHoverService(in args);
+    }
+
+    private readonly IClientCapabilitiesService _clientCapabilitiesService = args.ExportProvider.GetExportedValue<IClientCapabilitiesService>();
+
+    protected override IDocumentPositionInfoStrategy DocumentPositionInfoStrategy => PreferAttributeNameDocumentPositionInfoStrategy.Instance;
+
+    public ValueTask<RemoteResponse<Hover?>> GetHoverAsync(
+        JsonSerializableRazorPinnedSolutionInfoWrapper solutionInfo,
+        JsonSerializableDocumentId documentId,
+        Position position,
+        CancellationToken cancellationToken)
+        => RunServiceAsync(
+            solutionInfo,
+            documentId,
+            context => GetHoverAsync(context, position, cancellationToken),
+            cancellationToken);
+
+    private async ValueTask<RemoteResponse<Hover?>> GetHoverAsync(
+        RemoteDocumentContext context,
+        Position position,
+        CancellationToken cancellationToken)
+    {
+        var codeDocument = await context.GetCodeDocumentAsync(cancellationToken).ConfigureAwait(false);
+
+        if (!codeDocument.Source.Text.TryGetAbsoluteIndex(position, out var hostDocumentIndex))
+        {
+            return NoFurtherHandling;
+        }
+
+        var clientCapabilities = _clientCapabilitiesService.ClientCapabilities;
+        var positionInfo = GetPositionInfo(codeDocument, hostDocumentIndex, preferCSharpOverHtml: true);
+
+        if (positionInfo.LanguageKind == RazorLanguageKind.CSharp)
+        {
+            var generatedDocument = await context.Snapshot
+                .GetGeneratedDocumentAsync(cancellationToken)
+                .ConfigureAwait(false);
+
+            var csharpHover = await ExternalHandlers.Hover
+                .GetHoverAsync(
+                    generatedDocument,
+                    positionInfo.Position.ToLinePosition(),
+                    clientCapabilities.SupportsVisualStudioExtensions(),
+                    clientCapabilities.SupportsMarkdown(),
+                    cancellationToken)
+                .ConfigureAwait(false);
+
+            // Roslyn couldn't provide a hover, so we're done.
+            if (csharpHover is null)
+            {
+                return NoFurtherHandling;
+            }
+
+            // Map the hover range back to the host document
+            if (csharpHover.Range is { } range &&
+                DocumentMappingService.TryMapToHostDocumentRange(codeDocument.GetCSharpDocument(), range.ToLinePositionSpan(), out var hostDocumentSpan))
+            {
+                csharpHover.Range = RoslynLspFactory.CreateRange(hostDocumentSpan);
+            }
+
+            return Results(csharpHover);
+        }
+
+        if (positionInfo.LanguageKind is RazorLanguageKind.Html or RazorLanguageKind.Razor)
```

Comment:
This seems unnecssary. Perhaps it should be `is not` and throw unreachable? Or a debug assert?

Reply:
Derp - yes! You're right of course. :smile: